### ISSUE TITLE
Call with WebRTC support

### ISSUE THREAD

**Issue by bmarty** [vector-im/riotX-android#611]
### Introduction
WebRTC implementation supports voice and video calls in direct rooms which have only 2 joined members.

### Requirements

- [x] 1. Show voice and video call icons as menu actions in RoomDetailFragment only if the room is direct and has 2 joined members.

- [x] 2. Start a foreground service for both incoming and outgoing calls.

- [x] 3. Show heads-up notification for incoming calls with userId of the caller as title, call type as description, "Answer" and "Reject" buttons as actions.

- [x] 4. Show heads-up notification for outgoing calls with userId of the callee as title and "Calling..." as description.

- [x] 5. Change the title of the notifications as "Ongoing Call" as soon as the connection established.

- [x] 6. Notification (or CallActivity) should be visible even on the lock screen. 

- [x] 7. Show CallActivity when the outgoing call started and incoming call answered.

- [x] 8. Ongoing call has to continue even if CallActivity is destroyed by the user.

- [x] 9. Show CallActivity with the representation of the current call state when the user clicks the notification.

- [x] 10. Voice Call: Show hangup, mute, speaker buttons and the avatar of the participant on CallActivity.

- [x] 11. Video Call: Show hangup, mute, switch camera, turn off camera buttons and camera streams of both other participant and the current user on CallActivity.

- [x] 12. Automatically send hangup (busy) to the second incoming call if there is an ongoing call.

- [x] 13. Play sounds for ringing, ring back and hangup events.

- [x] 14. Send local ice candidate events to the room.

- [x] 15. Refresh turn server periodically.

- [x] 16. RiotX should not ring when performing an initial sync. It should handle still valid incoming call (regarding age and lifetime) properly though

- [x] 17. RiotX should behaves correctly if the incoming call is answered on another device.

- [x] 18. RiotX should ask for permission BEFORE starting on outgoing call (it's a bug currently). For incoming call, if the user denies the permission request, it should reject the call.

- [x] 19. `m.call.candidates` events should be hidden by default in the timeline

**Comment by RuralYak** [comment]
This is a showstopper for switching from Riot-android

**Comment by electricOzone** [comment]
> This is a showstopper for switching from Riot-android

I agree; could we get the appropriate showstopper tag added here?

**Comment by T-bond** [comment]
I also agree. The only reason keeping Riot on my phone to be able to manage calls/video conferences.

**Comment by 532910** [comment]
It should use [TelecomManager](https://developer.android.com/reference/android/telecom/TelecomManager) if I understood everything correctly.

**Comment by rajil** [comment]
For video conference, can we get an option to specify our own Jitsi server instead of hardcoding https://jitsi.riot.im/

**Comment by Thatoo** [comment]
> For video conference, can we get an option to specify our own Jitsi server instead of hardcoding https://jitsi.riot.im/

I believe that 1 to 1 calling is not using jitsi, it's webrtc peer to peer

**Comment by rajil** [comment]
@Thatoo agreed. I hope both p2p and Jitsi will land on RiotX. For Jitsi, the riot-web [already](https://matrix.org/blog/2020/04/06/running-your-own-secure-communication-service-with-matrix-and-jitsi) allows one to use their own self-hosted jitsi server. The mobile support is being requested here.

**Comment by KuenzelIT** [comment]
> @Thatoo agreed. I hope both p2p and Jitsi will land on RiotX. For Jitsi, the riot-web [already](https://matrix.org/blog/2020/04/06/running-your-own-secure-communication-service-with-matrix-and-jitsi) allows one to use their own self-hosted jitsi server. The mobile support is being requested here.

This would be very nice since it currently prevents our company from using matrix/riot.

**Comment by almereyda** [comment]
I also understand this issue to cover the WebRTC peer-to-peer calling within Riot.

There's an independent issue for Jitsi integration at #43

**Comment by floflodu34** [comment]
> using

I personnaly have a Matrix serveur and would like to configure jistsi server on app toi pls

**Comment by schiessle** [comment]
Especially in the current situation where people work more from home as before and can't meet their relatives in real life I think this becomes a really important feature. Because RiotX is otherwise already so much better then the "old" Android app I really don't want to switch back just to have the video/audio call functionality.

Maybe it would make sense to re-priorities this feature because of the special situation we are living?

**Comment by Dirmin** [comment]
It would be great to develop this feature as a priority, especially in this period of containment where video calls are increasingly used...
It would be a good idea if it could support at least 4 users before switching to a Jitsi solution!

**Comment by MatMaul** [comment]
I began porting the WebRTC code from Riot not X 2 weeks ago. It is not finished at all but since there is some activity around it I am pushing it here so the job doesn't get done twice.

This was done in Java so it is easy to diff with the original code for now, it would probably be nice to autoconvert that to Kotlin before the final merge.

https://github.com/MatMaul/riotX-android/tree/webrtc-wip

**Comment by Dirmin** [comment]
Amazing, thank you very much for this work !
This will undoubtedly be a much-used feature! How many people will be able to be on the same call?

**Comment by 532910** [comment]
@MatMaul, and all who are involved. Please implement personal calls with  [TelecomManager](https://developer.android.com/reference/android/telecom/TelecomManager). The same as [Koler](https://github.com/Chooloo/call_manage) does.

**Comment by nadonomy** [comment]
@niquewoodhouse this is just for 1:1 calls. For now I wouldn't concentrate on how to place the calls themselves, but would focus on the following:

### Calls
- Consider metadata to show during the call; the other persons avatar/name/user id, call duration etc
- Consider extra states (e.g. while waiting for the other user to answer)
- Button to switch camera
- Mute mic on/off toggle
- Video on/off toggle
- Change audio source? (Could load a bottom sheet with a list to select from?)
- Switch between SD/HD, perhaps behind some kind of contextual interaction
- A back/minimise button to hide the call and go back to Riot

### Resume call
- Use a [Resume call] toast if a call is ongoing but minimised
- Consider useful metadata to expose
- Consider placement, e.g. is it OK to occlude the main navigation

### Legacy devices
- We should be using the OS UI for answering calls (https://developer.android.com/guide/topics/connectivity/telecom/selfManaged)
- But, legacy devices may not support it, in which case we should use toasts with [Answer with voice] [Answer with video] & [Reject] or similar

**Comment by 532910** [comment]
Please, I'm begging you, the calls MUST be able to take and release with a bluetooth headset button!

**Comment by 532910** [comment]
> A back/minimise button to hide the call and go back to Riot

call should be spawned in a separate window, system back should bring you to the original riot screen, no separate back button is needed

**Comment by Graphicscore** [comment]
I would be really happy if this feature lands in RiotX soon, I feel like the current version of RiotX already is far more responsive and awesome to use than the current Riot Version. I would love to migrate as soon as possible.

**Comment by abdullasharapov** [comment]
RiotX is now better than Riot. Really lacking is just the ability to call. Please make every effort in this direction, I consider one of the most necessary functions

**Comment by sruggier** [comment]
I don't speak for this project, but as a general point of etiquette, I'd suggest you use the thumbs up reaction to vote for the issue instead of leaving comments, which spam every other person who is interested in following an issue. Developers can sort issues based on which ones had the most reactions, so it's a much more useful and friendly way to show your support.

If you're demanding specific features that aren't already implemented in Riot, it would be better to file those as separate issues. I think having feature parity sooner is more important than implementing new stuff on top while we all wait. Also, remember that if you haven't paid anyone to hack on this project, all you can do is politely make suggestions, you're not entitled to outright demand anything except a full refund of your payment, or lack thereof. :wink:

**Comment by 532910** [comment]
On the one hand I completely agree with @sruggier, on the other --- I fully understand @abdullasharapov. I think that the main matrix ecosystem strength is that it's not only a protocol but also a reference implementation of it. That's why jabber never will be usable --- there are no server and clients for all main platforms with the same set of supported extensions. In other words, I believe that keeping the entire ecosystem in sync is one of the top priorities. And as I see it is, except of riot-android/riotX, that I hope will catch it up soon.

**Comment by Kommynct** [comment]
Should there really be a button to go back to riot?

Why hide riot in the first place just for a call? I understand for video, but why bother doing that instead of just putting the buttons at the bottom over where you message others?

**Comment by 532910** [comment]
12. What about call holding and merging?

**Comment by notramo** [comment]
> Should there really be a button to go back to riot?
> 
> Why hide riot in the first place just for a call? I understand for video, but why bother doing that instead of just putting the buttons at the bottom over where you message others?
> 
> Working IT I have found many people are confused by calls in that they think their smartphone can't do both a call and something else at the same time, because of this antiquated way of doing things.

For voice call without video, it would be practical to do it similarly as Riot Web/Desktop. It should show a panel with hangup, speaker, and mic mute buttons, sticking to the top of the composer. The chat timeline and composer should be available without pressing any buttons. The "ongoing call" bar should be visible everywhere (e.g. sticking to the top/bottom of screen when viewing contact list, or app settings).

**Comment by 532910** [comment]
There should be an audio source seletion button instead of the just speaker button.

**Comment by BillCarsonFr** [comment]
> @MatMaul, and all who are involved. Please implement personal calls with [TelecomManager](https://developer.android.com/reference/android/telecom/TelecomManager). The same as [Koler](https://github.com/Chooloo/call_manage) does.

It will be done in a second PR, because we have to do it anyway for devices with no or limited support of telecom manager

**Comment by rajil** [comment]
The screen should get unlocked when a call is received. This is what Whatsapp does and gives user a better experience. Related bug [report](https://github.com/vector-im/riot-android/issues/3491).

**Comment by BillCarsonFr** [comment]
fixed by #1416

**Comment by 532910** [comment]
> > @MatMaul, and all who are involved. Please implement personal calls with [TelecomManager](https://developer.android.com/reference/android/telecom/TelecomManager). The same as [Koler](https://github.com/Chooloo/call_manage) does.
> 
> It will be done in a second PR, because we have to do without it anyway for devices with no or limited support of telecom manager

Should a separate issue be opened for this?
